packages/xpath/src/functions/xforms/node-set.ts

```diff
@@ -32,6 +34,221 @@ export const countNonEmpty = new NumberFunction(
+		// First argument is `target` (per spec) of the deepest resolved repeat
+		const target = args[0]!;
+
+		let pairs: EvaluatedIndexedRepeatArgumentPair[] = [];
+
+		// Iterate through rest of arguments, collecting pairs of:
+		//
+		// - `repeats`: **all** nodes matching the supplied node-set for the
+		//   `repeatN` argument in this pair
+		// - `position`: the resolved number value for the `indexN` (per spec)
+		//   argument at in this pair
+		//
+		// For **all `repeatN`/`indexN` pairs**, arguments are evaluated in the
+		// calling context (in typical XForms usage, this will be the context of the
+		// bound node). This is the core difference between this approach and the
+		// original in https://github.com/getodk/web-forms/pull/150. That
+		// understanding was clarified in review of that orignal effort, and is
+		// borne out by new tests exercising depth > 1, which demonstrate the same
+		// behavior in JavaRosa.
+		//
+		// Note: we start iterating here at index 1 so assertions related to
+		// positional argument index are clear.
+		for (let i = 1; i < args.length; i += 2) {
+			const repeatsArg = args[i];
+			const positionArg = args[i + 1];
+
+			assertArgument(i, repeatsArg);
+			assertArgument(i + 1, positionArg);
+
+			// Evaluate position first, because...
+			const position = positionArg.evaluate(context).toNumber();
+
+			// ... if any "index" (position) is `NaN`, we short-circuit. This is
+			// expected behavior because the equivalent `/data/repN[posN]/target`
+			// expression would do the same.
+			if (Number.isNaN(position)) {
+				return [];
+			}
+
+			// Reiterating the point made describing this loop for future clarity:
+			// this collects **all** of the nodes matching the `repeatN` expression.
+			// We filter them in a later step.
+			const repeats = evaluateArgumentToFilterableNodes(context, repeatsArg);
+
+			// No repeats = nothing to "index" = short circuit
+			if (repeats.length === 0) {
+				return [];
+			}
+
+			pairs.push({
+				repeats,
+				position,
+			});
+		}
+
+		// Sort the results of each `repeatN`/`indexN` pair, by containment order.
+		//
+		// Note: the `repeatN`/`indexN` pairs can be supplied in any order (this is
+		// consistent with behavior in JavaRosa, likely as a side effect of the
+		// function being implemented there by transforming the expression to its
+		// LocationPath equivalent).
+		pairs = pairs.sort(compareContainmentDepth);
+
+		// Resolve repeats at the specified/evaluated position, in document depth
+		// order by:
+		//
+		// 1. Filtering each set of repeats to include **only** the nodes contained
+		//    by the previously resolved repeat.
+		//
+		// 2. Selecting the repeat at the specified/evaluated position (of those
+		//    filtered in 1).
+		let repeatContextNode: ContextNode;
+
+		for (const pair of pairs) {
+			const { position } = pair;
+			const repeats = pair.repeats.filter((repeat) => {
+				return repeatContextNode?.contains(repeat) !== false;
```

Comment:
No, but I can see why it would seem that way! The reasoning for this was also less clear to me reading it today than when I wrote it.

It's a special case for the first iteration of the loop. I hope [7742cb9](https://github.com/getodk/web-forms/pull/195/commits/7742cb95f1570315ea8b84fd02430edab120c80c) is clarifying.

Reply:
Helpful for the first iteration, yes! What happens if I do `indexed-repeat(/some/repeat, /some/other_repeat, 3)`?

Reply:
I'm about to wrap up for the day so I can test tomorrow to say for sure, but I believe it will produce an empty node-set/blank string.

Ah, and now I understand the question better!

The `contains` filters aren't meant (only) as a guard against invalid structural references. They are also a result filter to roughly achieve the semantics of the the non-shorthand's LocationPath Step equivalent—i.e. `/data/subtree/leaf` produces a context of the `leaf` nodes contained by (well, in this case children of) `/data/subtree`.

You can think of it as conceptually similar to how some custom CSS selector engines work: they'll often work right to left (well, hierarchically bottom up), matching the deepest part of the selector first and then working hierarchically up the selector to refine the result set from there. (As described here, it's usually a performance optimization technique; and it's one we might also want to think about someday)

But yes, it will _also filter out_ invalid hierarchies as in your example, that's just not all it's doing.

Reply:
Makes sense, thanks! How about tracking the unrelated reference case as part of future error work? It would be nice to give explicit feedback as some sort of error if possible.